client-apis/decrypt.md

```diff
@@ -69,65 +91,226 @@ This default CMM MUST obtain the decryption materials required for decryption.
-specified by the algorithm suite.  
-The input to the KDF algorithm is the plaintext data key.
+This operation MUST perform these steps in order.
 
-The AAD used in decryption is the [Message Body AAD](../data-format/message-body-aad.md), constructed as follows:
+If the caller is streaming the encrypted plaintext to this operation
+and indicates an end to the encrypted message such that this operation can
+not complete the above steps,
+this operation MUST halt and indicate a failure to the caller.
 
-- Message ID: This value is the same as the message ID in the parsed message header.
-- Body AAD Content: This value depends on whether the encrypted content being decrypted is within a [regular frame](../data-format/message-body.md#regular-frame) ,
-  a [final frame](../data-format/message-body.md#final-frame) or is [non framed](../data-format/message-body.md#non-framed-data).
-  Refer to [Message Body AAD](../data-format/message-body-aad.md) specification for more information.
-- Sequence Number: This value is the [sequence number](../data-format/message-body.md#sequence-number) of the frame being decrypted, if the message contains framed data.
-  If the message contains non framed data, then this value is 1.
-- Content Length: TODO
+The caller MAY forgoe indicating an end to the encrypted message.
+This operation MAY still succeed as long as it can successfully process the first
+sequential bytes inputted to it as an encrypted message.
 
-If the algorithm suite has a signature algorithm, decrypt MUST verify the message footer using the specified signature algorithm,
-by using the verification key obtained from the decryption materials.
+If any of these steps fails, this oepration MUST falt and indicate a failure to the caller.
 
-## Security Considerations
+### Parse the header
+
+Given encrypted message bytes, this operation MUST process those bytes sequentially,
+deserializing those bytes according to the [message format](../data-format/message.md).
+
+This operation MUST attempt to deserialize all available encrypted message bytes until it has
+successfully deserialized a valid [message header](../data-format/message-header.md).
+
+This operation MUST wait if it doesn't have enough encrypted message bytes available to
+deserialize the next field of the message header until enough input bytes become available or
+the caller indicates an end to the encrypted message.
 
-[TODO]
+Once a valid message header is deserialized,
+this operation MUST validate the [message header body](../data-format/message-header.md#header-body)
+by using the [authenticated encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm)
+to decrypt with the following inputs:
 
-## Apendix
+- the AAD is the serialized [message header body](../data-format/message-header.md#header-body).
+- the IV has a value of 0
+- the cipherkey is the derived data key
+- the ciphertext is an empty byte array
```

Comment:
Maybe a bigger change, but this definition may better belong in the algorithm suite section rather than here. While all current suites use this technique, future ones may not.